src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,303 @@
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        task->bcast_linear.stage = STAGE_SETUP;
+    case STAGE_SETUP:
+        st = ucc_tl_cuda_bcast_linear_setup_test(task);
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        ucc_tl_cuda_put_sync(task);
+        if (trank == task->bcast_linear.root) {
+            task->bcast_linear.stage = STAGE_COPY;
+        } else {
+            task->bcast_linear.stage = STAGE_WAIT_ROOT;
+        }
+    default:
+        break;
+    }
+
+    if (trank == task->bcast_linear.root) {
+        // Root scenario
+        // fall-through between cases is intentional
+        switch (task->bcast_linear.stage) {
+        case STAGE_COPY:
+            // copy from src buffer to scratch
+            dbuf = PTR_OFFSET(TASK_SCRATCH(task, trank),
+                              task->bcast_linear.step % 2 * half_scratch_size);
+            sbuf = PTR_OFFSET(task->bcast_linear.sbuf, offset_buff);
+            st   = ecopy(dbuf, sbuf, chunk_size, exec,
+                       &task->bcast_linear.exec_task);
+            if (st != UCC_OK) {
+                ucc_error("failed to post ecopy task");
+                task->super.status = st;
+                return;
+            }
+            task->bcast_linear.stage = STAGE_WAIT_COPY;
+        case STAGE_WAIT_COPY:
+            etask = task->bcast_linear.exec_task;
+            if (etask) {
+                st = ucc_ee_executor_task_test(etask);
+                if (st == UCC_OK) {
+                    ucc_ee_executor_task_finalize(etask);
+                    task->bcast_linear.exec_task = NULL;
+                    // signal others
+                    ++task->bcast_linear.step;
+                    set_rank_step(task, task->bcast_linear.root,
+                                  task->bcast_linear.step, 0);
+                    task->bcast_linear.stage = STAGE_WAIT_ALL;
+                } else {
+                    // not ready
+                    return;
+                }
+            } else {
+                ucc_debug("etask is nullptr");
+                return;
+            }
+        case STAGE_WAIT_ALL:
+            for (i = 0; i < tsize; ++i) {
+                // need to wait until all ranks complete step - 1, because of double buffering
+                if (get_rank_step(task, i, 0) < task->bcast_linear.step - 1) {
+                    // rank is not ready, lets wait
+                    return;
+                }
+            }
+            task->bcast_linear.stage = STAGE_COPY;
+            if (task->bcast_linear.step < task->bcast_linear.num_steps) {
+                // go to next iteration
+                task->bcast_linear.stage = STAGE_COPY;
+                return;
+            } else {
+                // finish
+                task->bcast_linear.stage = STAGE_DONE;
+            }
+        case STAGE_DONE:
```

Comment:
could we remove this stage?

Reply:
yes, we could, I left it to be more explicit